plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+                    val index = (symbol.owner as IrValueParameter).index
+                    if (index >= 0) { // index == -1 for `this` parameter
+                        val transformedValueParameter = (parentDeclaration as IrFunction).valueParameters[index]
+                        return buildGetValue(transformedValueParameter.symbol)
+                    }
+                }
+            }
+            is IrCall -> {
+                dispatchReceiver?.let { dispatchReceiver = it.transformAtomicFunctionCall(parentDeclaration, lambda) }
+                extensionReceiver?.let { extensionReceiver = it.transformAtomicFunctionCall(parentDeclaration, lambda) }
+                getValueArguments().forEachIndexed { i, arg ->
+                    putValueArgument(i, arg?.transformAtomicFunctionCall(parentDeclaration, lambda))
+                }
+                val isInline = symbol.owner.isInline
+                val receiver = extensionReceiver ?: dispatchReceiver ?: return this
+                // Invocation of atomic functions on atomic receivers
+                // are substituted with the corresponding inline declarations from `kotlinx-atomicfu-runtime`
+                if (symbol.isKotlinxAtomicfuPackage() && receiver.type.isAtomicValueType()) {
+                    // 1. Receiver is the atomic field:
+                    // a.incrementAndGet() -> atomicfu_incrementAndGet(get_a {..}, set_a {..})
+                    if (receiver is IrCall) { // property accessor <get-field>
+                        val accessors = receiver.getAccessors(lambda ?: parentDeclaration)
+                        return inlineAtomicFunction(receiver.type.atomicToValueType(), accessors).apply {
+                            transformAtomicfuLoopBody(parentDeclaration)
+                        }
+                    }
+                    // 2. Atomic extension receiver `this`, the reciver of the parent function.
+                    // Parent function is the inline Atomic* extension function already transformed with `transformAtomicInlineDeclaration`
+                    // inline fun foo(getter: () -> T, setter: (T) -> Unit) { incrementAndGet() } ->
+                    // inline fun foo(getter: () -> T, setter: (T) -> Unit) { atomicfu_incrementAndGet(getter, setter) }
+                    if (receiver is IrGetValue && parentDeclaration.isTransformedAtomicExtensionFunction()) {
+                        val accessorParameters = (parentDeclaration as IrFunction).valueParameters.takeLast(2).map { it.capture() }
+                        return inlineAtomicFunction(receiver.type.atomicToValueType(), accessorParameters).apply {
+                            transformAtomicfuLoopBody(parentDeclaration)
+                        }
+                    }
+                } else {
+                    // 3. transform invocation of an inline Atomic* extension function call: a.foo()
+                    if (isInline && receiver is IrCall && receiver.type.isAtomicValueType()) {
```

Comment:
What if it call of custom extension function with `AtomicValue` extension receiver type? 
```
inline fun AtomicInt.foo() {  ... }
```

Reply:
I think it is worth adding some description of what transformation is doing here.